src/filter/bilateral.rs

```diff
@@ -0,0 +1,303 @@
+    fn weighted_average(weights_and_values: impl Iterator<Item = (f32, Self)>) -> Self {
+        let (weights_sum, weighted_values_sum) = weights_and_values
+            .map(|(w, v)| (w, w * f32::from(v.0[0])))
+            .reduce(|(w1, v1), (w2, v2)| (w1 + w2, v1 + v2))
+            .expect("cannot find a weighted average given no weights and values");
+
+        let weighted_average = weighted_values_sum / weights_sum;
+
+        Luma([weighted_average.round() as u8])
+    }
+}
+impl WeightedAverage for Rgb<u8> {
+    fn weighted_average(weights_and_values: impl Iterator<Item = (f32, Self)>) -> Self {
+        let (weights_sum, (r_weighed_sum, g_weighed_sum, b_weighted_sum)) = weights_and_values
+            .map(|(w, v)| {
+                (
+                    w,
+                    (
+                        (w * f32::from(v.0[0])),
+                        (w * f32::from(v.0[1])),
+                        (w * f32::from(v.0[2])),
+                    ),
+                )
+            })
+            .reduce(|(w1, (r1, g1, b1)), (w2, (r2, g2, b2))| (w1 + w2, (r1 + r2, g1 + g2, b1 + b2)))
+            .expect("cannot find a weighted average given no weights and values");
+
+        let r_average = r_weighed_sum / weights_sum;
+        let g_average = g_weighed_sum / weights_sum;
+        let b_average = b_weighted_sum / weights_sum;
+
+        Rgb([
+            r_average.round() as u8,
+            g_average.round() as u8,
+            b_average.round() as u8,
+        ])
+    }
+}
+
+/// Un-normalized Gaussian Weight
+fn gaussian_weight(x: f32, sigma: f32) -> f32 {
+    (-0.5 * x.powi(2) / sigma.powi(2)).exp()
+}
+
+/// A trait which provides a similarity metric between two pixels based on their colors.
+///
+/// This trait is used with the `bilateral_filter()` function.
+pub trait ColorSimilarity<P> {
+    /// Returns a measure of similarity between the two pixels based on their colors
+    fn color_similarity(&self, pixel1: &P, pixel2: &P) -> f32;
+}
+
+/// A gaussian function of the euclidean distance between two pixel's colors.
+///
+/// This implements [`ColorSimilarity`].
+pub struct GaussianEuclideanColorSimilarity {
+    /// The standard deviation of the gaussian distribution used on
+    /// the euclidean distance between pixels.
+    pub sigma: f32,
+}
+
+impl<P> ColorSimilarity<P> for GaussianEuclideanColorSimilarity
+where
+    P: Pixel,
+    f32: From<P::Subpixel>,
+{
+    fn color_similarity(&self, pixel1: &P, pixel2: &P) -> f32 {
+        //there is no need for sqrt'ing when a pixel has only one channel
+        let euclidean_distance = if P::CHANNEL_COUNT == 1 {
+            f32::from(pixel1.channels()[0]) - f32::from(pixel2.channels()[0]).abs()
+        } else {
+            pixel1
+                .channels()
+                .iter()
+                .zip(pixel2.channels().iter())
+                .map(|(c1, c2)| (f32::from(*c1) - f32::from(*c2)).powi(2))
+                .sum::<f32>()
+                .sqrt()
```

Comment:
This sqrt is immediately undone by the squaring within `gaussian_weight` - we should be able to get a bit of performance improvement by avoiding this unnecessary work.

Reply:
Good catch!